File: solutions/401.binary-watch/binary-watch.java
class Solution {
    public List<String> readBinaryWatch(int num) {
        List<String> res = new ArrayList<String>();
        int[] timeCode = new int[10];
        dfs(timeCode, 0, 0, num, res);
        return res;
    }
    private void dfs(int[] timeCode, int i, int k, int num, List<String> res) {
        if(k == num) {
            String curRes = deCodeTime(timeCode);
            if(curRes != null)
                res.add(curRes);
            return;
        }
        if(i == timeCode.length) return;
        timeCode[i] = 1;
        dfs(timeCode, i + 1, k + 1, num, res);
        timeCode[i] = 0;
        dfs(timeCode, i + 1, k, num, res);
    }
    private String deCodeTime(int[] timeCode) {
        int hours = 0;
        for(int i = 0; i < 4; i++) 
            if(timeCode[i] == 1)
                hours += Math.pow(2, i);
        int minutes = 0;
        for(int i = 4; i < 10; i++) 
            if(timeCode[i] == 1)
                minutes += Math.pow(2, i - 4);
        if(hours >= 12 || minutes >= 60) return null;
        return hours + (minutes < 10 ? ":0" : ":") + minutes;
    }
}

File: solutions/401.binary-watch/binary-watch.cpp
class Solution {
public:
    vector<string> readBinaryWatch(int num) {
        vector<string> times;
        for(int i = 0; i < 12; i++) {
            bitset<4> hours((size_t) i);
            for(int j = 0; j < 60; j++) {
                bitset<6> minutes((size_t) j);
                if(hours.count() + minutes.count() == num) 
                    times.push_back(to_string(i) + (j < 10 ? ":0" : ":") + to_string(j));
            }
        }
        return times;
    }
};